### PR TITLE
FIX: Destroy all posts when hard deleting topic

### PR DESCRIPTION
**Pull request by nbianca**
Hard deleting topics that contained soft deleted posts or small actions
used to create orphan posts because only the first post was hard
deleted. This commit adds an error message if there are still posts left
in the topic that must be hard deleted first or hard deletes all small
actions too immediately (there is no other way of hard deleting a small
action because there is no wrench menu).

### CODE REVIEW

File: app/controllers/topics_controller.rb [hunk 1]
```diff
@@ -622,19 +622,22 @@ def bookmark
...
-    force_destroy = false
     if params[:force_destroy].present?
       if !guardian.can_permanently_delete?(topic)
         return render_json_error topic.cannot_permanently_delete_reason(current_user), status: 403
       end
 
-      force_destroy = true
+      # can_permanently_delete? returns true even if there are still some
+      # small actions in the topic and these must be deleted too.
+      topic.posts.with_deleted.order(post_number: :desc).find_each do |post|
```
[the remark this answers is not in the document]
Comment (ZogStriP) [reply app/controllers/topics_controller.rb:632]: The controller method should be as simple as

```rails
def destroy
  topic = Topic.with_deleted.find_by(id: params[:id])
  first_post = topic.posts.with_deleted.order(:post_number).first

  if params[:force_destroy] == "true"
    guardian.ensure_can_permanently_delete!(topic)
    PostDestroyer.new(current_user, first_post, context: params[:context], force_destroy: true).destroy
  else
    guardian.ensure_can_delete!(topic)
    PostDestroyer.new(current_user, first_post, context: params[:context]).destroy
  end

  render body: nil
rescue Discourse::InvalidAccess
  render_json_error I18n.t("delete_topic_failed")
end
```

File: app/controllers/topics_controller.rb [hunk 2]
```diff
@@ -626,19 +626,17 @@ def destroy
...
-
-      first_post = topic.posts.with_deleted.order(:post_number).first
-      PostDestroyer.new(current_user, first_post, context: params[:context]).destroy
     end
 
+    PostDestroyer.new(
+      current_user,
+      topic.ordered_posts.with_deleted.first,
+      context: params[:context],
+      force_destroy: params[:force_destroy].present?
```
Comment (ZogStriP) [inline app/controllers/topics_controller.rb:637]: Will this work if we have this as a query string: `&force_destroy=false`?

File: app/controllers/topics_controller.rb [hunk 3]
```diff
@@ -622,19 +622,20 @@ def bookmark
...
       guardian.ensure_can_delete!(topic)
     end
 
-    first_post = topic.posts.with_deleted.order(:post_number).first
-    PostDestroyer.new(current_user, first_post, context: params[:context], force_destroy: force_destroy).destroy
+    PostDestroyer.new(
+      current_user,
+      topic.ordered_posts.with_deleted.first,
+      context: params[:context],
+      force_destroy: params[:force_destroy].present?
```
Comment (romanrizzi) [inline app/controllers/topics_controller.rb:637]: Can `params[:force_destroy]` could be `"false"` here? That would evaluate to true. I wonder if we should use `ActiveRecord::Type::Boolean.new.deserialize` here.

File: app/models/topic.rb [hunk 4]
```diff
@@ -1802,7 +1802,12 @@ def rate_limit_topic_invitation(invited_by)
   end
 
   def cannot_permanently_delete_reason(user)
-    if self.posts_count > 0
+    deleted_posts_count = Post.with_deleted
```
Comment (romanrizzi) [inline app/models/topic.rb:1805]: Are we missing a `.where.not(deleted_at: nil)` here? Do you think it's worth it to extract this into a scope?

Comment (ZogStriP) [reply app/models/topic.rb:1805]: That's exactly what the `with_deleted` scope does 😉

Comment (ZogStriP) [reply app/models/topic.rb:1805]: Actually, @romanrizzi is right, `Post.with_deleted` will be giving all posts (including those that are deleted), not just the ones that are deleted as the name of the variable suggests (`deleted_posts_count`).

Comment (ZogStriP) [reply app/models/topic.rb:1805]: Maybe just rename the variable to `with_deleted_posts_count`?

File: app/models/topic.rb [hunk 5]
```diff
@@ -1802,7 +1802,12 @@ def rate_limit_topic_invitation(invited_by)
...
+      .where(topic_id: self.id)
+      .where(post_type: [Post.types[:regular], Post.types[:moderator_action], Post.types[:whisper]])
+      .count
+
+    if posts_count > 0 || deleted_posts_count > 1
```
Comment (ZogStriP) [inline app/models/topic.rb:1810]: Why the `> 1` for `deleted_posts_count`?

File: app/controllers/topics_controller.rb (hunk 3 above)
Comment (nbianca) [reply app/controllers/topics_controller.rb]: Our UI does not ever set force_destroy to false and `ActiveRecord::Type::Boolean.new` has some issues too, but I made the change.

File: app/models/topic.rb (hunk 4 above)
Comment (nbianca) [reply app/models/topic.rb]: I renamed it to `all_posts_count`. :+1:

File: app/models/topic.rb (hunk 5 above)
Comment (nbianca) [reply app/models/topic.rb]: The current post is included in the count for `deleted_posts_count`. If there is another soft-deleted post in the topic, then the first post cannot be deleted.

**Review by ZogStriP — APPROVED** [review]

**Pull request merged**